jq/jq2.go

```diff
@@ -0,0 +1,493 @@
+	}
+
+	panic("unreachable")
+}
+
+const (
+	// KindNull represents the JSON value "null".
+	KindNull ValueKind = iota
+	// KindFalse represents the JSON value "false".
+	KindFalse
+	// KindTrue represents the JSON value "true".
+	KindTrue
+	// KindNumber represents the JSON type "number".
+	KindNumber
+	// KindString represents the JSON type "string".
+	KindString
+	// KindArray represents the JSON type "array".
+	KindArray
+	// KindObject represents the JSON type "object".
+	KindObject
+)
+
+func kind(jv C.jv) ValueKind {
+	switch C.jv_get_kind(jv) {
+	case C.JV_KIND_NULL:
+		return KindNull
+	case C.JV_KIND_FALSE:
+		return KindFalse
+	case C.JV_KIND_TRUE:
+		return KindTrue
+	case C.JV_KIND_NUMBER:
+		return KindNumber
+	case C.JV_KIND_STRING:
+		return KindString
+	case C.JV_KIND_ARRAY:
+		return KindArray
+	case C.JV_KIND_OBJECT:
+		return KindObject
+	}
+
+	panic("unreachable")
+}
+
+func jvToGoValue(jv C.jv) interface{} {
+	kind := C.jv_get_kind(jv)
+	switch kind {
+	case C.JV_KIND_NULL:
+		return nil
+	case C.JV_KIND_FALSE:
+		return false
+	case C.JV_KIND_TRUE:
+		return true
+	case C.JV_KIND_NUMBER:
+		val := C.jv_number_value(jv)
+		if C.jv_is_integer(jv) == 1 {
+			return int(val)
+		}
+		return float64(val)
+	case C.JV_KIND_STRING:
+		return C.GoString(C.jv_string_value(jv))
+	case C.JV_KIND_ARRAY:
+		arrayLen := int(C.jv_array_length(C.jv_copy(jv)))
+		array := make([]interface{}, arrayLen)
+		for i := 0; i < arrayLen; i++ {
+			v := C.jv_array_get(C.jv_copy(jv), C.int(i))
+			array[i] = jvToGoValue(v)
+			C.jv_free(v)
+		}
+		return array
+	case C.JV_KIND_OBJECT:
+		obj := make(map[string]interface{})
+		for iter := C.jv_object_iter(jv); C.jv_object_iter_valid(jv, iter) == 1; iter = C.jv_object_iter_next(jv, iter) {
+			jvKey := C.jv_object_iter_key(jv, iter)
+			jvValue := C.jv_object_iter_value(jv, iter)
+			// This is safe because jv_object_iter_key asserts the jv is a string.
+			key := C.GoString(C.jv_string_value(jvKey))
+
+			obj[key] = jvToGoValue(jvValue)
+
+			C.jv_free(jvKey)
+			C.jv_free(jvValue)
+		}
+		return obj
+	}
+	panic("unreachable")
+}
+
+func jvToValue(jv C.jv) Value {
+	return Value{kind(jv), jvToGoValue(jv)}
+}
+
+// Value represents a JSON value.
+type Value struct {
+	Kind    ValueKind
+	GoValue interface{}
+}
+
+func (v Value) Dump() string {
+	jv := v.jv()
+	defer C.jv_free(jv)
+	return dumpJvToGoStr(jv)
+}
+
+func dumpJvToGoStr(jv C.jv) string {
+	// jv_dump_string frees the provided jv, so we copy it.
+	dumpedjv := C.jv_dump_string(C.jv_copy(jv), C.int(0))
+	defer C.jv_free(dumpedjv)
+
+	// Strings from jv_string_value are cleaned up when the jv is freed.
+	return C.GoString(C.jv_string_value(dumpedjv))
+}
+
+func (v Value) String() string {
+	switch v.Kind {
+	case KindNull:
+		return "null"
+	case KindFalse:
+		return "false"
+	case KindTrue:
+		return "true"
+	}
+
+	return fmt.Sprintf("%v", v.GoValue)
+}
+
+func JSONValue(buf []byte) (Value, error) {
+	var i map[string]interface{}
```

Comment:
I believe this is wrong. A json value can be any json primitive (string, bool, etc)